### ISSUE TITLE
Mac OS X: Mouse Scrolling triggers command history instead of scrolling the window

### ISSUE THREAD

**Issue by dwt** [bpython/bpython#517]
Hi there,

I have to say that this is a bit annoying, as bpython (the shell variant) is one of the very few applications that redefine scroll wheel events in this way.

I am quite sure that this is a setting that can be changed - but the point of this bug report is that the default for that setting should really be to let the scroll wheel, well, scroll the window.

Instead what happens is that I scroll through the history of the commands I have invoked previously. True, that can also be interesting, but I am already perfectly capable of doing that with the up down arrows on my keyboard, which I need to use anyway when I want to execute commands.

So please change the default for the scroll wheel to actually scrolling the window.

**Comment by thomasballinger** [comment]
I can't replicate this - what terminal are you using? Is this inside of a terminal multiplexer like screen or tmux perhaps? When you say "the shell variant", which executable are you using, and which version of bpython?

**Comment by dwt** [comment]
Sorry, for more detail: This is the standard terminal on os X, no terminal multiplexer, zsh, bpython (without extra arguments).

```
$ bpython --version
bpython version 0.13.2 on top of Python 2.7.6
```

Do you need more information?

**Comment by dwt** [comment]
I've created a short movie to see the issue in action, as it seems I can't upload it to github, I uploaded it to  youtube https://www.youtube.com/watch?v=IHo-UVa8k8k

**Comment by thomasballinger** [comment]
The video seems to be private, I can't view it.

**Comment by dwt** [comment]
Sorry, it seems I upload too rarely. It should now be fixed. :)

**Comment by thomasballinger** [comment]
I still can't replicate this, the scroll wheel seems to scroll into the scrollback buffer in Terminal.app for me. Does the scroll wheel work for you when you use the terminal for other things? Is it possible you have something special set up make the scroll wheel work in vim or similar?

**Comment by dwt** [comment]
> I still can't replicate this, the scroll wheel seems to scroll into the scrollback buffer in Terminal.app for me. Does the scroll wheel work for you when you use the terminal for other things? Is it possible you have something special set up make the scroll wheel work in vim or similar?

I may have in LESS - but other than that my setup is normal - and outside of bpython scrolling works normally. Do you have any Idea what kind of setting could influence this or how the libraries you use for input are configured?

I would be fine to add some logging to the code to figure out what kind of input signal it receives if you can give me a pointer where to look in the code.

Best Regards,
Martin

**Comment by adamzap** [comment]
@dwt I think the issue is with your Terminal.app settings, this one in particular:

![](http://i.imgur.com/UZIGOtu.png)

The default is the behavior that you desire, and you can hold `Shift` to temporarily test the opposite behavior.

I hope you don't mind me jumping in on this issue, @thomasballinger

**Comment by thomasballinger** [comment]
Not at all, thanks @adamzap!

If this is indeed the case @dwt, then there's nothing I'd want to change - we probably can't tell the different between up arrow and the character the scroll wheel sends. I can't tell what bytes it sends because my terminal.app doesn't have this option - try running the program below and using your scroll wheel a bit during the blank screen:

~~~.sh
tput smcup; tmp=$(python -c 'import sys; print repr(sys.stdin.read(20))'); tput rmcup; echo $tmp
~~~
I'm guessing that the bytes displayed will be the same as using the arrow keys, so we won't be able to differentiate - but if they're different it's possible we could. All we could do in bpython though would be to trigger the pager or ignore them.

Therefore I'd further suggest that you update bpython - in the old version you're using you may be disappointed with scrolling behavior even if this were fixed, but in the new bpython scrolling works great! (and it doesn't use the alternate screen, so this won't be an issue)

**Comment by thomasballinger** [comment]
I'm going to assume @adamzap was correct and there's nothing to change here.

**Comment by dwt** [comment]
yes, the new version does fix the behaviour for me. Many thanks!

**Comment by gpinnaan** [comment]
At least in El Capitan and probably also in Yosemite you can move from scrolling the history to scrolling the window with cmd R

**Comment by ayatollah** [comment]
Sorry for waking up an old thread. 
I have the same problem as @dwt. However, I've found out that this only occures after I've been using ssh and timed out from the session, and only for that specifical window. If I open up a new windows, it works ok.

**Comment by lmichael107** [comment]
I never changed my settings, and just noticed this issue today after many months of expected behavior (scrolling terminal output rather than through command history). So, at some point, my settings were changed (to select the "Scroll alternate screen" box).

**Comment by Kornel** [comment]
@ayatollah Exactly this! I've just had that issue. A ssh session (with a screen running if that matters) terminated on my machine and there I was - scrolling through commands instead of the buffer :)

**Comment by dvmorris** [comment]
I have been experiencing the same issues with iTerm, but it does not seem to be related to bpython. Has anyone found a resolution? It doesn't happen all the time, and the only way I can get rid of the issue is by quitting iTerm and re-launching.

**Comment by danieljamesscott** [comment]
I'm having the same issue - ssh session timeout means that scrolling the window cycles the command history instead of scrolling the window

**Comment by shabadoo75** [comment]
This happens to me every now and then and I've not been able to track down the cause. As happened to @Kornel I did have an ssh session terminate on me immediately before it happend this time

**Comment by 1superk1** [comment]
Typing `reset` at the command prompt solves the issue....

**Comment by thomasballinger** [comment]
If anyone can replicate this issue (in bpython, although replicating with a disconnected ssh session could be useful background) I'd love to take a look. If this is fixable with reset then it's some terminal control sequence that bpython is emitting that it shouldn't be.

Let me know what OS, which terminal emulator, and how you can get it to occur.

Alternatively, if we knew what control sequence causes this behavior I could look for that in the code and prevent it. However it's hard to be confident in changes we can't test.

**Comment by 1superk1** [comment]
We're talking about Apple's Terminal. I keep having this issue since Snow
Leopard (OSX 10.6), while using bash...
The current system is OSX 10.12 Sierra, where this still occurs.

'reset' is a terminfo command as far as I know.... Has nothing to do with
bpython though. I think that both bash and bpython use readline, maybe
readline is the weak link.

There are two main issues that we experience in the case of a disconnected
ssh session: (1) if we ran 'screen' on the remote machine, the terminal
will not scroll anymore and the command prompt will span a single line at
the bottom of  Terminal's window and (2) the mouse wheel scrolling problem.

Both issues are resolved with a 'reset'



On Fri, Dec 16, 2016 at 12:21 AM, Thomas Ballinger <notifications@github.com

**Comment by thomasballinger** [comment]
So long as a disconnected ssh session is the only cause of this, then I agree it has nothing to do with bpython. While bpython doesn't use readline, it does spit out all kinds of terminal control sequences and tinker with termios/ioctl settings, which are the things that `reset` resets, hence my concern about it. If anyone is ever able to replicate causing this problem with bpython I'd be interested to hear.

But since is sounds like this is just a terminal thing unrelated to bpython, I'll keep the issue closed but you're welcome to keep discussing here :)

**Comment by dhoffi** [comment]
![image](https://user-images.githubusercontent.com/10828920/28117235-d5ce4f1e-670c-11e7-8937-edfa06c94dd8.png)

**Comment by dlamblin** [comment]
@dhoffi Is that the MacOS Terminal preferences? How did you get all these options instead of General, Profiles, Window Groups, and Encodings?

Anyway, I replicate this by using less in ssh, then using less on a file with more lines that the window height, then close the ssh connection (timeout, network issue, or just <enter>~.<enter>)
From that point you can use the touchpad scroll to go through the history only.
Less must have setup the alternate screen and scrolling of it.
Command-R will switch you back to the other scrolling, but so will issuing the command `reset`

It would not be such an annoying problem to encounter if the GUI had any other way to bring up the scroll bar so it could be clicked on other than to start scrolling (which it won't do when it's scrolling the history instead). Suddenly you have to know something specific about the terminal app, E.G. to use command-R, to scroll by page with shift-pgup or to reset the terminal state.

**Comment by Horsty** [comment]
@dlamblin You can see this screen if you go to preferences in iTerm. Not in the normal terminal.

**Comment by caruccio** [comment]
`CMD+,` -> `Advanced` -> Search `scroll`

Thanks A LOT @dhoffi

**Comment by Manojkumar18** [comment]
I use Linux for the same problem too I fix it,

![scrolling](https://user-images.githubusercontent.com/31440315/44560973-ecbfc780-a76f-11e8-9fd9-c0653cc15f71.png)

click enable -> Scroll on Output.

**Comment by 26piyushkedia26** [comment]
Just open a new terminal. This will solve your problem.

**Comment by ldo4jnpr** [comment]
I found the following which explains what is causing this issue: https://support.apple.com/guide/terminal/display-or-hide-the-alternate-screen-trmld1f46097/mac. You can select View -> Hide Alternate Screen and mouse scrolling will work again.

**Comment by renecouto** [comment]
> Typing `reset` at the command prompt solves the issue....

this also clears all the past output that I wanted to search :upside_down_face:

**Comment by surak** [comment]
If someone has this problem in specific terminals, function-scroll always work

**Comment by divStar** [comment]
Yeah, fn+scroll scrolls the buffer, just scroll (e.g. using two fingers or mouse wheel) scrolls command history **in the original terminal** on MacOS 14+ (just tested this a second ago).

**Comment by surak** [comment]
> Yeah, fn+scroll scrolls the buffer, just scroll (e.g. using two fingers or mouse wheel) scrolls command history **in the original terminal** on MacOS 14+ (just tested this a second ago).

Not by default, no

**Comment by divStar** [comment]
> Not by default, no

Actually that's correct. I had been using a `screen` session, which is why the regular swipe / scroll did not work, but fn+scroll did. I don't have any custom shortcuts set up.

**Comment by dlamblin** [comment]
Okay so:
> Suddenly you have to know something specific about the terminal app [in this case iTerm], E.G. to use command-R, to scroll by page with shift-pgup or to reset the terminal state. *…*

…or to hold <kbd>fn</kbd> while scrolling. Bear in mind, a number of external keyboards do not have the same <kbd>fn</kbd> key to hold. I often have my macbook on an adjustable arm making its internal keyboard not something I'll reach for easily.

It might be good to show the scrollbar interactive elements in both regular and alternate mode, it could also be handy to show Alt-Screen/Scroll is enabled somewhere in the scroll overlay (like find), title, status line (if on).

**Comment by i1t** [comment]
TL;DR: the solution from this thread: https://github.com/bpython/bpython/issues/517#issuecomment-93853607
